core/base/src/main/java/alluxio/collections/ConcurrentIdentityHashMap.java

```diff
@@ -0,0 +1,347 @@
+        .collect(Collectors.toSet());
+  }
+
+  @Override
+  public void putAll(Map<? extends K, ? extends V> map) {
+    map.forEach((k, v) -> mInternalMap.put(new IdentityObject<>(k), v));
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return mInternalMap.isEmpty();
+  }
+
+  @Override
+  public int size() {
+    return mInternalMap.size();
+  }
+
+  @Override
+  public boolean containsValue(Object value) {
+    return mInternalMap.containsValue(value);
+  }
+
+  /**
+   * Retrieves a set representing all keys contained within the map at time of calling.
+   *
+   * Note this method departs slightly from the standard {@link ConcurrentHashMap} implementation
+   * by providing its own implementation of a Set for {@link IdentityObject}. Not all methods
+   * have been implemented. Namely the {@code toArray} and {@code add*}.
+   *
+   * @return The set of keys in the map.
+   */
+  @Override
+  public Set<K> keySet() {
+    Set<IdentityObject<K>> set = mInternalMap.keySet();
+    return new Set<K>() {
+      @Override
+      public int size() {
+        return set.size();
+      }
+
+      @Override
+      public boolean isEmpty() {
+        return set.isEmpty();
+      }
+
+      @Override
+      public boolean contains(Object o) {
+        return set.contains(new IdentityObject<>(o));
+      }
+
+      @Override
+      public Iterator<K> iterator() {
+        return set.stream().map(IdentityObject::get).iterator();
+      }
+
+      @Override
+      public Object[] toArray() {
+        throw new UnsupportedOperationException("toArray not supported in identity keySet");
+      }
+
+      @Override
+      public <T> T[] toArray(T[] a) {
+        throw new UnsupportedOperationException("toArray not supported in identity keySet");
+      }
+
+      @Override
+      public boolean add(K k) {
+        throw new UnsupportedOperationException("add not supported in keySet");
+      }
+
+      @Override
+      public boolean remove(Object o) {
+        return set.remove(new IdentityObject<>(o));
+      }
+
+      @Override
+      public boolean containsAll(Collection<?> c) {
+        return set.containsAll(c);
+      }
+
+      @Override
+      public boolean addAll(Collection<? extends K> c) {
+        throw new UnsupportedOperationException("addAll not supported in keySet");
+      }
+
+      @Override
+      public boolean retainAll(Collection<?> c) {
+        return set.retainAll(c);
+      }
+
+      @Override
+      public boolean removeAll(Collection<?> c) {
+        return set.retainAll(c);
+      }
+
+      @Override
+      public void clear() {
+        set.clear();
+      }
+    };
+  }
+
+
+  private class IdentityObject<T> {
+    private T mObj;
+
+    public IdentityObject(T o) {
+      this.mObj = o;
+    }
+
+    public T get() {
+      return mObj;
+    }
+
+    @Override
+    public int hashCode() {
+      // Robert Jenkins 32-bit integer hash
+      int a = System.identityHashCode(mObj);
+      a = (a+0x7ed55d16) + (a<<12);
```

Comment:
Why do we need to hash output of `System.identityHashCode()`?

Reply:
This is similar to the behavior that the netty version of this class implemented. I'm not 100% sure why, but having this here vs. not using the jenkins hash I think will depend on how much we trust the `System.identityHash` to be a "good" hash.

Reply:
Here's the relevant code: 

https://github.com/netty/netty/blob/3.10/src/main/java/org/jboss/netty/util/internal/ConcurrentIdentityHashMap.java#L112-L142

Reply:
I see. In a nutshell this is to regularize bit positions of hash value for efficient bucketing. Lets mention the reason in comments.